### PR TITLE
Provide metadata about the origin of the devfile with DevWorkspace

### PR DESCRIPTION
**Pull request by olexii4**
Signed-off-by: Oleksii Orel <oorel@redhat.com>

### What does this PR do?
This PR provide adding a metadata about the origin of the devfile with DevWorkspace(it depends on the case):

1. From devfile location (https://gist.githubusercontent.com/l0rd/71a04dd0d8c8e921b16ba2690f7d5a47/raw/d520086e148c359b18c229328824dfefcf85e5ef/spring-petclinic-devfile-v2.0.0.yaml)
```
annotations:
  che.eclipse.org/devfile-source: |
      url:
        location:  >-https://gist.githubusercontent.com/l0rd/71a04dd0d8c8e921b16ba2690f7d5a47/raw/d520086e148c359b18c229328824dfefcf85e5ef/spring-petclinic-devfile-v2.0.0.yaml
```

2. From project location (https://github.com/l0rd/spring-petclinic/tree/bfbad6a85aee019018e9c67bcec7ff759af7de1d)
```
annotations:
  che.eclipse.org/devfile-source: |
      scm:
        repo: 'https://github.com/l0rd/spring-petclinic.git'
        revision: bfbad6a85aee019018e9c67bcec7ff759af7de1d
        fileName: devfile.yaml
```

3. From one of the sample (Java Spring Boot)
```
annotations:
  che.eclipse.org/devfile-source: |
     sample:
          registry: https://che-eclipse-che.apps.cluster-28e4.28e4.sandbox1899.opentlc.com/devfile-registry
          displayName: Java Spring Boot
          location: 'https://github.com/che-samples/java-spring-petclinic/tree/devfilev2'
```

4. From custom devfile
```
annotations:
  che.eclipse.org/devfile-source: |
      custom: {}
```

You can tested these changes with  https://che-eclipse-che.apps.cluster-28e4.28e4.sandbox1899.opentlc.com

### What issues does this PR fix or reference?
fixes https://github.com/eclipse/che/issues/19430
fixes https://github.com/eclipse/che/issues/19967

### CODE REVIEW

**Comment by olexii4** [comment]
[dashboard-ci-test]

File: src/pages/GetStarted/CustomWorkspaceTab/DevfileSelector/index.tsx [hunk 1]
```diff
@@ -79,9 +81,15 @@ export class DevfileSelectorFormGroup extends React.PureComponent<Props, State>
   private async handleDevfileSelect(meta: che.DevfileMetaData): Promise<void> {
     // clear location input
     this.devfileLocationRef.current?.clearInput();
+    const cheDevworkspaceEnabled = this.props.workspacesSettings['che.devworkspaces.enabled'] === 'true';
+    let devfile: api.che.workspace.devfile.Devfile;
     try {
-      const devfileContent = await this.props.requestDevfile(meta.links.self) as string;
-      const devfile = safeLoad(devfileContent);
+      if (cheDevworkspaceEnabled) {
```
Comment (sleshchenko) [inline src/pages/GetStarted/CustomWorkspaceTab/DevfileSelector/index.tsx:87]: I think if statement can be moved out of try

File: src/pages/GetStarted/__tests__/updateDevfileMetadata.spec.ts [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+  });
+
+  it('should return a devfile V2 with a sample metadata', () => {
+    const metadata = getMetadata();
+    const attributes = {
+      'dw.metadata.annotations': {
+        'che.eclipse.org/devfile-source': safeDump({
+          sample: {
+            registry: metadata.registry,
+            name: metadata.displayName,
```
Comment (sleshchenko) [inline src/pages/GetStarted/__tests__/updateDevfileMetadata.spec.ts:56]: well, PR description has name (`name: 'apache-camel-springboot'` which is not display name), but here in tests you use display name as name, it's confusing

File: src/pages/GetStarted/__tests__/updateDevfileMetadata.spec.ts [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+    expect(devfile.metadata.attributes).toEqual(attributes);
+  });
+
+  it('should return a devfile V2 with a sample metadata', () => {
+    const metadata = getMetadata();
+    const attributes = {
+      'dw.metadata.annotations': {
+        'che.eclipse.org/devfile-source': safeDump({
+          sample: {
```
Comment (sleshchenko) [inline src/pages/GetStarted/__tests__/updateDevfileMetadata.spec.ts:54]: after I used sample on quick add page I got DevWorkspace created with scm source(which may makes even more sense than sample but still). Could you elaborate how to test sample devfile-source:
Screenshot_20210614_142327

Comment (olexii4) [reply src/pages/GetStarted/__tests__/updateDevfileMetadata.spec.ts:54]: I have updated the target image for https://che-eclipse-che.apps.cluster-28e4.28e4.sandbox1899.opentlc.com.
Fixed.

File: src/pages/GetStarted/__tests__/updateDevfileMetadata.spec.ts (hunk 2 above)
Comment (olexii4) [reply src/pages/GetStarted/__tests__/updateDevfileMetadata.spec.ts:56]: I have fixed it:
```name: metadata.displayName``` -> ```displayName: metadata.displayName```

File: src/pages/GetStarted/CustomWorkspaceTab/DevfileSelector/index.tsx (hunk 1 above)
Comment (olexii4) [reply src/pages/GetStarted/CustomWorkspaceTab/DevfileSelector/index.tsx:87]: I think it is not a good idea. Next line is a request:
```
await this.props.requestFactoryResolver(meta.links.v2);
```

Comment (olexii4) [reply src/pages/GetStarted/CustomWorkspaceTab/DevfileSelector/index.tsx:87]: We have a request which depends on boolean variable(one request or another).

Comment (sleshchenko) [reply src/pages/GetStarted/CustomWorkspaceTab/DevfileSelector/index.tsx:87]: you're right. My fault

**Review by akurinnoy — APPROVED** [review]
LGTM

**Comment by sleshchenko** [comment]
Tested and works as expected:

Sample:
```yaml
apiVersion: workspace.devfile.io/v1alpha2
kind: DevWorkspace
metadata:
  annotations:
    che.eclipse.org/devfile-source: |
      sample:
        registry: >-
          https://che-eclipse-che.apps.cluster-28e4.28e4.sandbox1899.opentlc.com/devfile-registry
        displayName: Java Spring Boot
        location: 'https://github.com/che-samples/java-spring-petclinic/tree/devfilev2'

```

Git repo
```yaml
apiVersion: workspace.devfile.io/v1alpha2
kind: DevWorkspace
metadata:
  annotations:
    che.eclipse.org/devfile-source: |
      scm:
        repo: 'https://github.com/che-samples/java-spring-petclinic.git'
        revision: devfilev2
        fileName: devfile.yaml
```

url
```yaml
apiVersion: workspace.devfile.io/v1alpha2
kind: DevWorkspace
metadata:
  annotations:
    che.eclipse.org/devfile-source: |
      url:
        location: >-
          https://raw.githubusercontent.com/che-samples/java-spring-petclinic/devfilev2/devfile.yaml
```

then cluster died and I was not able to test custom workspace flow but I believe it works as expected =)

**Review by sleshchenko — APPROVED** [review]

**Pull request merged**